Add database schema

kahakai opened the pull request:
- Add README.
- Add database schema image to README.
- Implement configuration from file & environment variables.
- Add timestamptz support.
- Add UserRepository class for database queries.
- Update User model.
- Update /users API endpoint.
- Update basic project structure.
- Upgrade core dependencies.

kahakai committed:
Implement configuration from file & environment variables

Clean up dependencies.

kahakai committed:
Implement POST /users endpoint

Add Users table schema with Exposed DSL.
Add User models.
Add UserRepository class for queries.

kahakai committed:
Add database schema image

vitalyraevsky commented on build.gradle.kts:
```diff
@@ -1,25 +1,24 @@
 plugins {
-    kotlin("jvm") version "1.9.22"
+    kotlin("jvm") version "1.9.23"
     id("io.ktor.plugin") version "2.3.8"
```
тогда уже и подними ktor до 2.3.9

vitalyraevsky commented on src/main/kotlin/com/ermolnik/api/details/DetailsApi.kt:
```diff
@@ -21,15 +18,15 @@ fun Application.detailsApi() {
                         "Боевик",
                         "22.04.2021",
-                        "$HOST_URL:$PORT/cover",
+                        "https://upload.wikimedia.org/wikipedia/ru/5/55/%D0%93%D0%BD%D0%B5%D0%B2_%D1%87%D0%B5%D0%BB%D0%BE%D0%B2%D0%B5%D1%87%D0%B5%D1%81%D0%BA%D0%B8%D0%B9_%28%D0%BF%D0%BE%D1%81%D1%82%D0%B5%D1%80%29.png",
```
TODO перевести на собственный CDN

kahakai replied:
Good suggestion, although all these stubs will be removed anyways.

vitalyraevsky commented on src/main/kotlin/com/ermolnik/api/users/User.kt:
```diff
@@ -0,0 +1,31 @@
+
+@Serializable
+data class CreateUserRequest(
+    @SerialName("username")
```
не избыточно ли SerialName , если филд и нейминг не отличается?

kahakai replied:
I prefer things to be explicit. Another point here is that we will still have to leave one field with `@SerialName` because it differs when serialized but every other one will be without `@SerialName`, which can make the situation confusing and misleading for some people because you will have to know and remember every time how this implicit way of names serialization works.

vitalyraevsky commented on src/main/kotlin/com/ermolnik/api/users/UserApi.kt:
```diff
@@ -0,0 +1,40 @@
+    routing {
+        // Create user
+        post("/users") {
+            val req = call.receive<CreateUserRequest>()
```
может скоуп функцию with?

kahakai replied:
Next PR

vitalyraevsky commented on src/main/kotlin/com/ermolnik/db/Database.kt:
```diff
@@ -0,0 +1,25 @@
+import org.jetbrains.exposed.sql.transactions.experimental.newSuspendedTransaction
+
+fun Application.configureDatabase(): Database {
+    val dbDriver = environment.config.property("ktor.db-driver").getString()
```
здесь тоже можно `with(environment.config)`

vitalyraevsky replied:
или приватный экст

kahakai replied:
Next PR

kahakai committed:
Bump Ktor version

kahakai committed:
Add timestamptz support

Add exposed-kotlin-datetime dependency.
Update Users table to include timestamptz columns.
Add OffsetDateTimeSerializer class to support timestamp serialization.
Update UserRepository for new User model.

ermolnik commented on src/main/kotlin/com/ermolnik/Application.kt:
```diff
@@ -1,22 +1,19 @@
-    embeddedServer(Netty, port = PORT, host = HOST_URL, module = Application::module)
-        .start(wait = true)
-}
+fun main(args: Array<String>) = io.ktor.server.netty.EngineMain.main(args)
```
а для чего ты поменял? в чем профит от нетти?

kahakai replied:
We're still using Netty here, the same server engine:
- `embeddedServer(Netty...`
+ `io.ktor.server.netty...`

The only change is the way of creating a server in our code, because embeddedServer is too constrained in terms of flexibility and customization and is suitable only for prototypes and rapid development when you pass all the parameters in code.
The main reason for change to `EngineMain` is the ability to use configuration file, we will need to be able to load a configuration file and environment variables for deployment anyways.

See https://ktor.io/docs/engines.html#choose-create-server
